File: ts/wasm/lib/detailed/complex_operations/account_update.ts
import { operation } from "../protocol.js";
import type { TAccountName } from "../hive_apps_operations/index.js";
import { OperationBase, IOperationSink } from "../operation_base.js";
import type { IHiveChainInterface } from "../interfaces.js";
import { HiveAccountCategory } from "./role_classes/categories/hive_authority/index.js";
import { RoleCategoryBase } from "./role_classes/role_category_base.js";
import { WaxError } from "../errors.js";

export type { THiveRoles, HiveAccountCategory, HiveRoleAuthorityDefinition, HiveRoleMemoKeyDefinition } from "./role_classes/categories/hive_authority/index.js";

// Here are all of the role categories. They are automatically parsed. Add new categories here
export const AuthorityRoleCategories = [
  HiveAccountCategory
] as const satisfies Readonly<Array<(new () => RoleCategoryBase<any>)>>;

export type UnionToIntersection<U> =
  (U extends any ? (k: U) => void : never) extends
  (k: infer I) => void ? I : never;

export type TRolesStruct = UnionToIntersection<InstanceType<typeof AuthorityRoleCategories[number]>["authorities"]>;
export type TRole = keyof TRolesStruct;

export type TRoleContainerNames = InstanceType<typeof AuthorityRoleCategories[number]>["category"];

export type TRoleKeyToValueMap = {
  [K in keyof TRolesStruct]: Omit<TRolesStruct[K], 'init' | 'enforceModifications'>
};

export type TRoleContainerKeyToValueMap = {
  [C in InstanceType<typeof AuthorityRoleCategories[number]> as C["category"]]: {
    [K in keyof C["authorities"]]: Omit<C["authorities"][K], 'init' | 'enforceModifications'>
  }[keyof C["authorities"]];
};

/**
 * Online operation - it is automatically filled in with the data acquired from the chain API.
 *
 * The purpose of this operation is to simplify account authority update process by automatic gathering current authority data
 * from blockchain and then supplementing them by provided action methods. Operation is designed to support different authority categories
 * (right now is implemented Hive builtin authority).
 *
 * It is initialized with all of the supported roles for the given account using {@link AccountAuthorityUpdateOperation.createFor} automatically, so
 * for example: If you want to add a single key to the active level of the account, you can just add the key, without worrying about your other key authorities.
 *
 * After initialization, you can use the {@link AccountAuthorityUpdateOperation.role} method to retrieve the role instance for the given role level.
 *
 * @example
 * ```ts
 * const operation = await AccountAuthorityUpdateOperation.createFor(myChain, "initminer");
 *
 * const active = operation.role("active");
 * active.add(myKey);
 *
 * const memo = operation.role("memo");
 * memo.set(myKey);
 * ```
 */
export class AccountAuthorityUpdateOperation extends OperationBase {
  private constructor(
    private readonly instancesPerContainerName: Map<TRoleContainerNames, RoleCategoryBase<any>>,
    private readonly instancesPerRoleName: Map<TRole, RoleCategoryBase<any>>
  ) {
    super();
  }

  /**
   * Creates an instance of AccountAuthorityUpdateOperation with all supported roles pre-initialized for the given account.
   *
   * @param {IHiveChainInterface} chain chain interface required for online user roles parsing
   * @param {TAccountName} account account we will operate on
   * @returns {Promise<AccountAuthorityUpdateOperation>} initialized account authority update operation
   */
  public static async createFor(chain: IHiveChainInterface, account: TAccountName): Promise<AccountAuthorityUpdateOperation> {
    const roles = new Map<TRole, RoleCategoryBase<any>>();
    const instances = new Map(AuthorityRoleCategories.map(role => {
      const container = new role();

      return [container.category, container];
    }));

    for (const roleClass of instances.values()) {
      await roleClass.init(chain, account);

      for (const roleName in roleClass.authorities)
        roles.set(roleName as TRole, roleClass);
    }

    return new AccountAuthorityUpdateOperation(instances, roles);
  }

  /**
   * Enforces the requirement for **owner** role authorization when modifying **active** or **posting** roles.
   * 
   * **HF 28** introduces stricter matching between the authority role required by a given operation and the role used to authorize the transaction.
   * - Since modifying **active** or **posting** roles requires **active** authority at the time of transaction signing, the pre-HF28 behavior — which allowed signing with the **owner** key — will be **disallowed**.  
   * - This change may pose difficulties for users who have lost their active keys and attempt to use their owner key to set a new one.  
   * - To address this, the function allows the inclusion — within the `account_update2_operation` generated internally — of elements that enforce the **owner** role requirement, specifically through an ineffective change of the owner authority to the same value currently recorded on-chain.
   */
  public enforceOwnerRoleAuthorisation(): void {
    const roleInstance = this.instancesPerRoleName.get("owner")!;
    const ownerRole = roleInstance.authorities["owner"];
    ownerRole.enforceModifications()
  }

  /**
   * Returns the role instance for the given role level.
   *
   * @param {KRole} level role level to retrieve
   * @returns Role class instance for the given role level
   */
  public role<KRole extends TRole>(level: KRole): TRoleKeyToValueMap[KRole] {
    const roleInstance = this.instancesPerRoleName.get(level);

    if (!roleInstance)
      throw new Error(`Role level ${level} is not initialized`);

    return roleInstance.authorities[level];
  }

  /**
   * Returns all of the roles for the given role category.
   *
   * Note: You can differentiate between different role categories by using the `level` property
   *
   * @param {KRoleContainer} category role category to retrieve
   * @returns Iterable of all roles for the given role category
   *
   * @example
   * ```ts
   * for(const role of operation.roles("hive"))
   *   if (role.level === "memo")
   *     role.set(myKey);
   *   else
   *     role.add(myKey);
   * ```
   */
  public roles<KRoleContainer extends keyof TRoleContainerKeyToValueMap>(category: KRoleContainer): Iterable<TRoleContainerKeyToValueMap[KRoleContainer]> {
    const roleInstance = this.instancesPerContainerName.get(category);

    if (!roleInstance)
      throw new Error(`Role category ${category} is not initialized`);

    return Object.values(roleInstance.authorities);
  }

  /**
   * Checks if the authority has changed since the last update and it is possible to transmit this operation (any changes applied)
   */
  public get isEffective(): boolean {
    let effective = false;

    for (const role of this.instancesPerContainerName.values())
      effective ||= role.changed;

    return effective;
  }

  /**
   * @internal
   */
  public finalize(sink: IOperationSink): Iterable<operation> {
    const operations: operation[] = [];

    for (const role of this.instancesPerContainerName.values())
      operations.push(...role.finalize(sink));

    if (operations.length === 0)
      throw new WaxError("No operations updating account authority generated");

    return operations;
  }
}

File: python/wax/complex_operations/account_update.py
from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import TYPE_CHECKING, Any, Iterable

from wax._private.operation_base import ConvertedToProtoOperation, OperationBase
from wax.complex_operations.data_providers.authority_data_provider import OnlineChainAuthorityDataProvider
from wax.complex_operations.role_classes.hive_authority.hive_account_category import (
    HiveAccountCategory,
)
from wax.complex_operations.role_classes.hive_authority.hive_roles import HiveRoles
from wax.exceptions.validation_errors import (
    NoAuthorityOperationGeneratedError,
)

if TYPE_CHECKING:
    from wax.complex_operations.role_classes.level_base import LevelBase
    from wax.complex_operations.role_classes.role_category_base import RoleCategoryBase
    from wax.interfaces import IAuthorityDataProvider, IHiveChainInterface, IWaxBaseInterface
    from wax.models.basic import AccountName


@dataclass
class PossibleRoles(HiveRoles):
    """All of possible roles. Add new roles to as base classes."""


AuthorityRoleCategories = (HiveAccountCategory,)
"""All of the role categories. Add new categories here."""


@dataclass
class PossibleCategoriesByCategoryName:
    """All of possible categories grouped by name. Add new categories also here."""

    hive: HiveAccountCategory


class AccountAuthorityUpdateOperation(OperationBase):
    """
    Account authority update operation.

    Notes:
        To create an instance of this class, use the factory method `create_for`.
        Otherwise, assertion error will be raised.
    """

    def __init__(
        self,
        possible_categories: PossibleCategoriesByCategoryName,
        possible_roles: PossibleRoles,
        *,
        _private: bool = False,
    ) -> None:
        assert (
            _private
        ), "AccountAuthorityUpdateOperation should be created only by AccountAuthorityUpdateOperation.create_for"
        super().__init__()
        self._possible_categories: PossibleCategoriesByCategoryName = possible_categories
        self._possible_roles: PossibleRoles = possible_roles

    @staticmethod
    async def create_for(
        wax_interface: IHiveChainInterface,
        account_name: AccountName,
    ) -> AccountAuthorityUpdateOperation:
        """
        Factory method for creating AccountAuthorityUpdateOperation.

        This is an online version - creates default provider instance and does call to api.

        Args:
            wax_interface (IHiveChainInterface): instance of interface.
            account_name (AccountName): account name of account to gain authority data from.

        Returns:
            AccountAuthorityUpdateOperation instance.
        """
        provider = OnlineChainAuthorityDataProvider(wax_interface)
        return await AccountAuthorityUpdateOperation.create_for_with_provider(wax_interface, account_name, provider)

    @staticmethod
    async def create_for_with_provider(
        wax_interface: IWaxBaseInterface, account_name: AccountName, provider: IAuthorityDataProvider
    ) -> AccountAuthorityUpdateOperation:
        """
        Factory method for creating AccountAuthorityUpdateOperation.

        This one support providing own authority data provider.

        Args:
            wax_interface (IWaxBaseInterface): instance of interface.
            account_name (AccountName): account name of account to gain authority data from.
            provider (IAuthorityDataProvider): authority provider.

        Returns:
            AccountAuthorityUpdateOperation instance.
        """
        role_per_role_name: dict[str, LevelBase[Any]] = {}
        category_per_category_name: dict[str, RoleCategoryBase[Any]] = {}

        for category in AuthorityRoleCategories:
            container = category()
            category_per_category_name[container.category] = container

        for category_instance in category_per_category_name.values():
            await category_instance.init(wax_interface, account_name, provider)

            for role_name, role_instance in category_instance.authorities.__dict__.items():
                role_per_role_name[role_name] = role_instance  # noqa: PERF403

        return AccountAuthorityUpdateOperation(
            PossibleCategoriesByCategoryName(**category_per_category_name),  # type: ignore[arg-type]
            PossibleRoles(**role_per_role_name),  # type: ignore[arg-type]
            _private=True,
        )

    def enforce_owner_role_authorisation(self) -> None:
        """
        Enforces the requirement for **owner** role authorization when modifying **active** or **posting** roles.

        **HF 28** introduces stricter matching between the authority role required by a given operation and the role
        used to authorize the transaction.
            - Since modifying **active** or **posting** roles requires **active** authority at the time of transaction
              signing, the pre-HF28 behavior — which allowed signing with the **owner** key — will be **disallowed**.
            - This change may pose difficulties for users who have lost their active keys and attempt to use their
              owner key to set a new one.
            - To address this, the function allows the inclusion — within the `account_update2_operation` generated
              internally — of elements that enforce the **owner** role requirement, specifically through an ineffective
              change of the owner authority to the same value currently recorded on-chain.
        """
        self._possible_roles.owner.enforce_modifications()

    @property
    def roles(self) -> PossibleRoles:
        """
        Get role to modify.

        Typical usage example:
            ```account_update_operation.roles.active.any_action()```
        """
        return self._possible_roles

    @property
    def categories(self) -> PossibleCategoriesByCategoryName:
        """Get category to modify."""
        return self._possible_categories

    @property
    def is_effective(self) -> bool:
        return any(category.changed for category in asdict(self.categories).values())

    def finalize(self, api: IWaxBaseInterface) -> Iterable[ConvertedToProtoOperation]:
        """
        Finalize updating account authority.

        Args:
            api (IWaxBaseInterface): Wax api subclass instance.

        Returns:
            Iterable[ConvertedToProtoOperation]: List of operations.

        Raises:
            NoAuthorityOperationGeneratedError: If no operation was generated.
        """
        operations: list[ConvertedToProtoOperation] = [
            operation for category in asdict(self.categories).values() for operation in category.finalize(api)
        ]

        if not operations:
            raise NoAuthorityOperationGeneratedError

        return operations
